lib/ABI/AddABIAwarenessToFunctionPrototypes.cpp

```diff
@@ -0,0 +1,210 @@
+    break;
+
+  case model::TypeKind::CABIFunctionType: {
+    Result = true;
+    using CABIFT = model::CABIFunctionType;
+    for (const auto &A : llvm::cast<CABIFT>(&Type)->Arguments())
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, A.Type());
+    const auto &ReturnType = llvm::cast<CABIFT>(&Type)->ReturnType();
+    Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, ReturnType);
+  } break;
+
+  case model::TypeKind::RawFunctionType: {
+    Result = true;
+    using RawFT = model::RawFunctionType;
+    for (const auto &A : llvm::cast<RawFT>(&Type)->Arguments()) {
+      auto Kind = model::Register::primitiveKind(A.Location());
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Kind);
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, A.Type());
+    }
+    for (const auto &V : llvm::cast<RawFT>(&Type)->ReturnValues()) {
+      auto Kind = model::Register::primitiveKind(V.Location());
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Kind);
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, V.Type());
+    }
+
+    const auto &Stack = llvm::cast<RawFT>(&Type)->StackArgumentsType();
+    revng_assert(Stack.Qualifiers().empty());
+    const model::TypePath &Internal = Stack.UnqualifiedType();
+    if (Internal.isValid())
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Internal);
+  } break;
+
+  default:
+    revng_abort("Unknown type.");
+  }
+
+  if (Result)
+    VH.setVerified(&Type);
+
+  VH.verificationCompleted(&Type);
+
+  rc_return VH.maybeFail(Result);
+}
+
+static bool needsVectorRegisterSupport(model::VerifyHelper &VH,
+                                       const model::TypePath &Type) {
+  return !needsVectorRegisterSupportImpl(VH, model::QualifiedType(Type, {}));
+}
+
+using namespace std::string_literals;
+
+class AddABIAwarenessToFunctionPrototypes {
+public:
+  static constexpr auto Name = "AddABIAwarenessToFunctionPrototypes";
+  inline static const std::tuple Options = { pipeline::Option("abi",
+                                                              "Invalid"s) };
+  std::vector<std::vector<pipeline::Kind *>> AcceptedKinds = {};
+
+  void run(pipeline::Context &Context, std::string InputABI) {
+    auto &Model = revng::getWritableModelFromContext(Context);
+
+    // Minimize the negative impact on binaries with ABI that is not fully
+    // supported by disabling the conversion by default.
+    //
+    // Use revng model pass on those functions to convert them manually.
+    //
+    // TODO: extend this list.
+    static constexpr std::array ABIsTheConversionIsEnabledFor = {
+      model::ABI::SystemV_x86_64,        model::ABI::SystemV_x86,
+      model::ABI::SystemV_x86_regparm_3, model::ABI::SystemV_x86_regparm_2,
+      model::ABI::SystemV_x86_regparm_1, model::ABI::AAPCS
+    };
+    if (!llvm::is_contained(ABIsTheConversionIsEnabledFor, Model->DefaultABI()))
+      return;
```

Comment:
* Why do we care about `DefaultABI` when we have an argument for this?
* Log a message

Reply:
1. Done: https://github.com/revng/revng/pull/232/commits/7525c68841b421bbb62fecf63706e35f1fe5765e.

Reply:
:+1: close if 2) is done too.

Reply:
That's why I didn't close it yet. There are other threads related to the analysis-related logger, so I'll catch this one when I deal with those.

Reply:
Done.